### ISSUE TITLE
Rethink the semantics of `require` and `import`

### ISSUE THREAD

**Issue by fatcerberus** [spheredev/neosphere#280]
`require` was implemented in early versions of miniSphere as the primary facility for using modules; it was never compatible with Node.js’s module search algorithm and in fact was never intended to be.  As a result it was superseded by ES module support in miniSphere 5.0.0 and now only exists as legacy.

There is no compatibility requirement for code written prior to the release of miniSphere 5, and `require` is not covered by Sphere v2 API versioning rules.

At present (miniSphere 5.2.14), calling `require()` and passing it the filename of an ES module will load the module synchronously and return its module namespace.  This is, in fact, a violation of the ECMAScript specification which requires that modules be loaded from the event loop (i.e. with no JS code on the stack).  It also means that `require`, despite existing primarily to load CommonJS modules, is subject to the same module resolution protocol as `import`, which means (as of the latest release), `nodeModule.js` will load as ESM rather than CommonJS.  Much like `import` falling back on a CJS shim for `.js` files (see #279), this violates the principle of least astonishment.  If `require` is going to exist at all, it should exist only for interop with Node.js and not try to be clever.

Long story short: `require` is almost certainly broken at present and needs to have its semantics rethought.

**Comment by fatcerberus** [comment]
I almost want to say that `require()` now is a similar case to `RequireScript()`: it exists only for the sake of loading CommonJS code and so it would make most sense to have it always load modules as CommonJS, regardless of the file extension, the same way `RequireScript()` always evaluates as traditional JS.

This does raise the issue of what to do about `import`, as it's still possible to load CommonJS modules this way.  Node.js with `--experimental-modules` can currently do the same, so I'm not convinced it makes sense to drop support for this.  I'll give it some more thought.

**Comment by fatcerberus** [comment]
Here's what I think I'll do: Starting with miniSphere 5.3, to test the waters, `require()` will be changed to always load whatever file is passed in as a CommonJS module, and `import` will likewise only support loading as ESM.  The main script for v2 games will always be loaded as ESM--never CommonJS.  `*.mjs` will also be removed from the search path for `require()`.

This would have the benefit that the type of code being loaded would be immediately obvious just from reading the call site.  `require` means it's CommonJS, and `import` means ESM, always.

**Comment by fatcerberus** [comment]
The compromise mentioned in #287 is no longer the case; `require()` in miniSphere 5.3 will only load either CommonJS modules or JSON files and *never* ES modules.  All ESM code must be loaded using `import` (either the static or dynamic variant) going forward and `require` remains only for compatibility with a subset of modules written for Node.js.

I consider this issue to be solved now, so I've changed the milestone to 5.3.0 and will now close it.
